trilitech/jstz
# feat(core,proto,kernel,cli): nested transactions

# Context




**Related Tasks**: [Support nested transactions in `jstz_proto`](https://app.asana.com/0/1205770721173531/1206345416213104/f) 

This PR adds the long awaited support for nested transactions in `jstz_proto`, permitting _atomic synchronous nested smart function calls_. 

# Description






Myself and @alanmarkoTrilitech discovered some issues with the old implementation of nested transactions -- namely the use of lifetimes caused issues if we used `Rc<RefCell<_>>` for a pointer to the parent transaction. 

There were two possible solutions to this:
- Wrapping all snapshot values in `Rc<RefCell<_>>`, thus eliminating most lifetimes from the transaction API
- Using the transactional stack design from https://www.notion.so/trilitech/Transactional-Storage-4c4a4b173145498dbbb61448e505ee16?pvs=4

This PR implements the latter. 

This PR contains:
- An entire rewrite of transactions to use the above design
- The integration of nested transactions with all JavaScript runtime APIs
- The integration of nested transactions with `jstz_proto` and `jstz_kernel`

# Manually testing the PR



To observe nested transactions:
```
cargo run --bin jstz -- deploy examples/nested_counter.js
```
Save the address `A`. Tail the function. 
Run
```
cargo run --bin jstz -- run "tezos://${A}/?incr=${A}"
```
